Rector resolving deprecations for Symfony 4.2

escopecz opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [Y]
| New feature/enhancement? (use the a.x branch)      | [Y]
| Deprecations?                          | [N]
| BC breaks? (use the c.x branch)        | [Y]
| Automated tests included?              | [Y, mostly existing already] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:

This PR is increasing Rector automatic check from Symfony 4.1 o 4.2 depreactions. The main topic was to stop getting dependencies from container. I was able to remove some places and ignored other places for various reasons that are documented in the rector config file.

The biggest refactoring in this PR is related to removing the `LegacyLeadModel` as that was also flagged by the Rector rule. So I rather removed this deprecated class completely. It was quite a rabbit hole as it forced me to move some code around to avoid circular dependencies. Luckily, we have quite a lot of test covering the contact merging so I'm quite confident this did not break anything.

I added one more test to increase code coverage for DWC route. It turned out this route was broken for some time as it was linked to method that did not exist. Or the name was a bit different. So I fixed that route but I'm not quite sure if it's useful when no one missed it.

I had to use different approach to load fixtures during installations. Loading them from a folder did not work as we to put dependencies to them now when we cannot do it via the container. The replacement was quite elegant actually as we can use the `SymfonyFixturesLoader` instead and use already existing group to load the fixtures we want and use DI. However this service is not mockable so the existing unit test was failing because of it. I had to create a wrapper class with interface to fix this design flaw in Doctrine's service.

#### Steps to test this PR:

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Try to merge contacts via API, import, UI, page hits

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

escopecz committed:
Letting Rector resolve deprecations for Symfony 4.2

escopecz committed:
Removing deprecated LegacyLeadModel finally so we don't have to maintain it.

It was quite a rabbit hole to remove it without DI loops so it needed some refactoring.

escopecz committed:
Reverting some changes Rector did and ignoring those with notes why

escopecz committed:
Removing entityManager property from AbstractMauticMigration to avoid getting it from the container

escopecz committed:
CS fixes

escopecz committed:
Using DI for some fixtures over container

escopecz committed:
CS, STAN fixes

escopecz committed:
Catching SameContactException when merging contacts

escopecz committed:
Fixing tests

escopecz committed:
New functional tests for the /dwc public route (which was broken not by this PR)

The route was looking for getEntitiesAction but only getAction existed

escopecz commented on app/bundles/DynamicContentBundle/Config/config.php:
```diff
@@ -27,7 +27,7 @@
             'mautic_api_dynamicContent_index' => [
                 'path'       => '/dwc',
-                'controller' => 'Mautic\DynamicContentBundle\Controller\DynamicContentApiController::getEntitiesAction',
+                'controller' => 'Mautic\DynamicContentBundle\Controller\DynamicContentApiController::getAction',
```
I'm not sure if this endpoint is used, because `getEntitiesAction` method does not exist in the controller. `getAction` method does exist. So it's a random fix here but I wrote tests for this and it showed this up.

escopecz committed:
Removing duplicated method expectation

escopecz committed:
Is this test breaking the others?

escopecz committed:
Yes, it was! How about disabling global state?

RCheesley commented:
When testing this PR on GitPod I am seeing the following error at install when it loads fixtures:

```
1.2 - Loading fixtures...
In LeadFieldData.php line 17:                                                                                                                                                
Too few arguments to function Mautic\InstallBundle\InstallFixtures\ORM\LeadFieldData::__construct(), 0 passed in /var/www/html/vendor/doctrine/data-fixtures/lib/Doctrine/Common/DataFixtures/Loader.php on line 231 and exactly 1 expected         
```

RCheesley commented:
Also this when trying to load Mautic after installation completes:

```
500 Internal Server Error - The "Mautic\InstallBundle\Controller\InstallController::getParameter()" method is missing a parameter bag to work properly. Did you forget to register your controller as a service subscriber? This can be fixed either by using autoconfiguration or by manually wiring a "parameter_bag" in the service locator passed to the controller. 

Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException
/vendor/symfony/framework-bundle/Controller/AbstractController.php:68 at
/app/bundles/InstallBundle/Controller/InstallController.php:221 at Symfony\Bundle\FrameworkBundle\Controller\AbstractController -> getParameter ( 'kernel.root_dir' )
/vendor/symfony/http-kernel/HttpKernel.php:159 at Mautic\InstallBundle\Controller\InstallController -> stepAction ( '0' )
/vendor/symfony/http-kernel/HttpKernel.php:81 at Symfony\Component\HttpKernel\HttpKernel -> handleRaw ( object(Request), '1' )
/vendor/symfony/http-kernel/Kernel.php:201 at Symfony\Component\HttpKernel\HttpKernel -> handle ( object(Request), '1', true )
/app/AppKernel.php:104 at Symfony\Component\HttpKernel\Kernel -> handle ( object(Request), '1', true )
/app/middlewares/CORSMiddleware.php:82 at AppKernel -> handle ( object(Request), '1', true )
/app/middlewares/CatchExceptionMiddleware.php:34 at Mautic\Middleware\CORSMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/Dev/IpRestrictMiddleware.php:52 at Mautic\Middleware\CatchExceptionMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/VersionCheckMiddleware.php:58 at Mautic\Middleware\Dev\IpRestrictMiddleware -> handle ( object(Request), '1', true )
/app/middlewares/TrustMiddleware.php:42 at Mautic\Middleware\VersionCheckMiddleware -> handle ( object(Request), '1', true )
/vendor/stack/builder/src/Stack/StackedHttpKernel.php:23 at Mautic\Middleware\TrustMiddleware -> handle ( object(Request), '1', true )
/vendor/stack/run/src/Stack/run.php:13 at Stack\StackedHttpKernel -> handle ( object(Request) )
/index_dev.php:25 at Stack\run ( object(StackedHttpKernel) )
```

mollux commented:
@escopecz @RCheesley 

I have the same issue

```
2022-10-07 13:01:58] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException: "The "Mautic\InstallBundle\Controller\InstallController::getParameter()" method is missing a parameter bag to work properly. Did you forget to register your controller as a service subscriber? This can be fixed either by using autoconfiguration or by manually wiring a "parameter_bag" in the service locator passed to the controller." at /var/www/html/mautic/vendor/symfony/framework-bundle/Controller/AbstractController.php line 68 {"exception":"[object] (Symfony\\Component\\DependencyInjection\\Exception\\ServiceNotFoundException(code: 0): The \"Mautic\\InstallBundle\\Controller\\InstallController::getParameter()\" method is missing a parameter bag to work properly. Did you forget to register your controller as a service subscriber? This can be fixed either by using autoconfiguration or by manually wiring a \"parameter_bag\" in the service locator passed to the controller. at /var/www/html/mautic/vendor/symfony/framework-bundle/Controller/AbstractController.php:68)"} {"hostname":"37146bea162e","pid":40}
```

I can't get through the installer so most likely an issue in the UI installer.

mollux commented:
@escopecz 5.x has the same issue, so a recent change broke the UI installer, I'm going after it

mollux commented:
@escopecz @RCheesley see https://github.com/mautic/mautic/pull/11553
Issue is not related to this PR. 
UI installer is broken since 23/09

RCheesley commented:
Thanks @mollux - tested your PR and it fixed the issue. Surprised this was not picked up by automated tests?

escopecz commented:
@RCheesley I rebased this for another round of tests. I have also another PR in works for Symfony 4.3 rule set: https://github.com/mautic/mautic/pull/11555

But this one must be merged first

RCheesley commented:
Just tried this one but Mautic did not install properly on Gitpod - just merged in the latest changes from the branch and will re-test.

escopecz commented:
There is something wrong with this. I run the API tests against this and it failed. https://github.com/mautic/api-library/actions/runs/3219149158/jobs/5264158477 I'll take a look at what's going on.

RCheesley commented:
@escopecz I am trying to test this but I am still seeing issues with the UI installer being shown when it completes, with the following error:

```
1.2 - Loading fixtures...
In LeadFieldData.php line 17:
Too few arguments to function Mautic\InstallBundle\InstallFixtures\ORM\LeadFieldData::__construct(), 0 passed in /var/www/html/vendor/doctrine/data-fixtures/lib/Doctrine/Common/DataFixtures/Loader  
  .php on line 231 and exactly 1 expected  
```

This was fixed by the PR which was merged earlier, so it should be included in this PR, but it's still throwing the same error.

Did something maybe go wrong with the rebase?

escopecz committed:
Fetching install fixtures from SymfonyFixturesLoader rather than container

as we need dependencies in the fixtures and loading them from a folder won't allow us to use dependencies

escopecz commented:
https://github.com/mautic/mautic/pull/11543/commits/7879570eb8e544ca95ef9a38169264deb30a5aa4 is fixing the error you mention. But I'm still fixing the CI-reported issues. I'll ping you once the CI is green

escopecz committed:
Resolving issue with unmockable Doctrine service

escopecz commented:
@RCheesley @mollux @kuzmany or @biozshock This is ready to test. It also [passes the API tests](https://github.com/mautic/api-library/actions/runs/3225480383) which it did not before.

RCheesley reviewed (approved):
Tested merging contacts using the importer, using the user interface and using the API which all worked perfectly.

escopecz committed:
Merge branch '5.x' into rector-up-to-symfony-4.2

escopecz commented:
I'm going to merge this as it's blocking https://github.com/mautic/mautic/pull/11555. I'd like to move forward as the deadline is getting closer.
